### DISCUSSION TITLE
execute script before poety build starts

### DISCUSSION THREAD

**Discussion by tvogel-nid** [python-poetry/poetry#7704] [General]
from what I read [here](https://github.com/python-poetry/poetry/issues/5539) it seems to be possible to have a script that runs before the actual build process starts.

So I created an entry 
```
[tool.poetry.build]
generate-setup-file = false
script = "pre_build_nivcl.py"
```
and a  script like:
```
def parse_args():
    argparser = argparse.ArgumentParser(add_help=False)
    argparser.add_argument('--assemblySourcePath', help='path to the dotnet assemblies after build/publish', default='../../../../Binaries/NetCore/Publish/', required=False)
    argparser.add_argument('--version', help='version number', required=False)
    args, unknown = argparser.parse_known_args()
    print("the inputs are:")
    for arg in vars(args):
        print("{} is {}".format(arg, getattr(args, arg)))

    return args

def build():
    """
    This function is mandatory in order to build the extensions.
    """
    inputs=parse_args()
    print("---------- Input parameters ---------")
    print(inputs)
    print("-------------------------------------")

if __name__ == "__main__":
    build()
```

but it seems not to be executed at all, I don't even get an error message.

Did I just understood something wrong?

**Comment by clintonroy** [comment]
Please always give the full configuration files and commands that you have run, otherwise it's nearly impossible for us to figure out what's going on.

At least for me, your script is being run enough that i get import errors:

```
lintonr@rassilon:~/projects/poet/poetryplay$ poetry build
Preparing build environment with build-system requirements poetry-core
Building aldotaibot (2.3)
Traceback (most recent call last):
  File "/home/clintonr/projects/poet/poetryplay/pre_build_nivcl.py", line 24, in <module>
    build()
  File "/home/clintonr/projects/poet/poetryplay/pre_build_nivcl.py", line 18, in build
    inputs=parse_args()
           ^^^^^^^^^^^^
  File "/home/clintonr/projects/poet/poetryplay/pre_build_nivcl.py", line 4, in parse_args
    argparser = argparse.ArgumentParser(add_help=False)
                ^^^^^^^^
NameError: name 'argparse' is not defined. Did you mean: 'argparser'?


```

**Reply by tvogel-nid** [reply]
Hi Clinton,

thanks a lot for the quick answer and sorry for the late response.
You are rigth that the script itself is not bugfree, I just wanted to be sure that it gets executed at all as I din't see any output at this point of time.
This problem seems to be solved, now I'm wondering whether it is possible to give dynamic parameters to the script?

Thanks a lot
Thomas

**Reply by tvogel-nid** [reply]
after some time I couldn't work on my project I restarted and finally figured out that the 
```
[tool.poetry.build]
generate-setup-file = false
script = "pre_build_nivcl.py"
```

is only used when just doing a `poetry build` but not if I use `poetry build --format=sdist`, in the later case the script doesn't get called.

Is this intended or why is this script only executed if I ceate a wheel?

**Reply by tvogel-nid** [reply]
and another question is, when exactly is this script executed, it seems it is only executed after the *.tar.gz and *.whl is created.
If this is true, is there a way to define the execution order, so execute script before the build of the package?
